src/lib/babylon/scene/logic/raycasts.ts

```diff
@@ -0,0 +1,197 @@
+import { PBRaycastResult } from "@dcl/protocol/out-ts/decentraland/sdk/components/raycast_result.gen";
+import { LastWriteWinElementSetComponentDefinition } from "../../../decentraland/crdt-internal/components";
+import { raycastComponent, raycastResultComponent } from "../../../decentraland/sdk-components/raycast-component";
+import { SceneContext } from "../context";
+import { globalCoordinatesToSceneCoordinates, sceneCoordinatesToBabylonGlobalCoordinates } from "../coordinates";
+import { BabylonEntity } from "../entity";
+
+/**
+ * The processRaycasts function iterates over a copy of the pendingRaycastOperations
+ * and for each it does
+ * 1. It performs the final ray transformations based on the final positions of the entities
+ * 2. Filters the meshes to perform the raycast
+ * 3. Updates the RaycastResult component with the result of the query
+ * 4. If necessary, removes the raycast from pendingRaycastOperations
+ */
+export function processRaycasts(scene: SceneContext) {
+  const RaycastResult = scene.components[raycastResultComponent.componentId] as LastWriteWinElementSetComponentDefinition<PBRaycastResult>
+
+  // clone the set into an array to mutate the set while iterating
+  const iter = Array.from(scene.pendingRaycastOperations)
+  for (const entityId of iter) {
+    const raycast = scene.components[raycastComponent.componentId].get(entityId) as PBRaycast | null
+
+    const shouldRaycastBeDeletedFromPendingList = !raycast?.continuous
+
+    if (raycast) {
+      const entity = scene.getEntityOrNull(entityId)
+      if (entity && entity.appliedComponents.raycast) {
+        const ray = computeRayDirection(scene, raycast, entity.appliedComponents.raycast.ray, entity)
+
+        // get a list of all possible meshes to project this ray to
+        const intersectableMeshes = Array.from(getMeshesWithMask(scene.rootNode, raycast.collisionMask ?? 0xffff_ffff))
+
+        // then perform the actual raycast
+        const results = ray.intersectsMeshes(intersectableMeshes, false)
+
+        // and start preparing the result
+        const raycastResult: PBRaycastResult = {
+          direction: ray.direction,
+          globalOrigin: globalCoordinatesToSceneCoordinates(scene, ray.origin),
+          timestamp: raycast.timestamp || 0,
+          hits: []
+        }
+
+        if (raycast.queryType === RaycastQueryType.RQT_HIT_FIRST && results.length) {
+          raycastResult.hits = [pickingToRaycastHit(scene, pickClosest(results)!, ray)]
+        } else if (raycast.queryType === RaycastQueryType.RQT_QUERY_ALL && results.length) {
+          raycastResult.hits = results.map(_ => pickingToRaycastHit(scene, _, ray))
+        }
+
+        // send the result back to the scene
+        RaycastResult.createOrReplace(entity.entityId, raycastResult)
+      }
+    }
+
+    // lastly remove the raycast from the list if necessary
+    if (shouldRaycastBeDeletedFromPendingList) {
+      scene.pendingRaycastOperations.delete(entityId)
+    }
```

Comment:
_Not-naming suggestion_
`shouldRaycastBeDeletedFromPendingList` is declared pretty far from its only use. I also don't feel it's necessary, for readability purposes at least. This is super clear without it too, maybe with a short comment:

```js
if (!raycast.continuous) {
  scene.pendingRaycastOperations.delete(entityId) // only process once
}
```

(Or `raycast?.continuous` if left outside the `if (raycast)`, another thing to consider)

Reply:
fixed